server/src/main/java/org/apache/iotdb/db/metadata/storagegroup/SGMManager.java

```diff
@@ -0,0 +1,1762 @@
+
+  private TimeseriesStatistics timeseriesStatistics = TimeseriesStatistics.getInstance();
+  private MTreeBelowSG mtree;
+  // device -> DeviceMNode
+  private LoadingCache<PartialPath, IMNode> mNodeCache;
+  private TagManager tagManager;
+
+  // region Interfaces and Implementation of initialization、snapshot、recover and clear
+  public SGMManager() {
+    mtreeSnapshotInterval = config.getMtreeSnapshotInterval();
+    mtreeSnapshotThresholdTime = config.getMtreeSnapshotThresholdTime() * 1000L;
+
+    int cacheSize = config.getmManagerCacheSize();
+    mNodeCache =
+        Caffeine.newBuilder()
+            .maximumSize(cacheSize)
+            .build(
+                new com.github.benmanes.caffeine.cache.CacheLoader<PartialPath, IMNode>() {
+                  @Override
+                  public @Nullable IMNode load(@NonNull PartialPath partialPath)
+                      throws MetadataException {
+
+                    return mtree.getNodeByPath(partialPath);
+                  }
+                });
+
+    if (config.isEnableMTreeSnapshot()) {
+      timedCreateMTreeSnapshotThread =
+          IoTDBThreadPoolFactory.newSingleThreadScheduledExecutor("timedCreateMTreeSnapshot");
+      timedCreateMTreeSnapshotThread.scheduleAtFixedRate(
+          this::checkMTreeModified,
+          MTREE_SNAPSHOT_THREAD_CHECK_TIME,
+          MTREE_SNAPSHOT_THREAD_CHECK_TIME,
+          TimeUnit.SECONDS);
+    }
+
+    if (config.getSyncMlogPeriodInMs() != 0) {
+      timedForceMLogThread =
+          IoTDBThreadPoolFactory.newSingleThreadScheduledExecutor("timedForceMLogThread");
+
+      timedForceMLogThread.scheduleAtFixedRate(
+          this::forceMlog,
+          config.getSyncMlogPeriodInMs(),
+          config.getSyncMlogPeriodInMs(),
+          TimeUnit.MILLISECONDS);
+    }
+  }
+
+  // Because the writer will be used later and should not be closed here.
+  @SuppressWarnings("squid:S2093")
+  public synchronized void init(IStorageGroupMNode storageGroupMNode) {
+    if (initialized) {
+      return;
+    }
+
+    storageGroupFullPath = storageGroupMNode.getFullPath();
+
+    sgSchemaDirPath = config.getSchemaDir() + File.separator + storageGroupFullPath;
+    File sgSchemaFolder = SystemFileFactory.INSTANCE.getFile(sgSchemaDirPath);
+    if (!sgSchemaFolder.exists()) {
+      if (sgSchemaFolder.mkdirs()) {
+        logger.info("create storage group schema folder {}", sgSchemaFolder.getAbsolutePath());
+      } else {
+        logger.info(
+            "create storage group schema folder {} failed.", sgSchemaFolder.getAbsolutePath());
+      }
+    }
+    logFilePath = sgSchemaDirPath + File.separator + MetadataConstant.METADATA_LOG;
+
+    logFile = SystemFileFactory.INSTANCE.getFile(logFilePath);
+
+    try {
+      // do not write log when recover
+      isRecovering = true;
+
+      tagManager = new TagManager(sgSchemaDirPath);
+      tagManager.init();
+      mtree = new MTreeBelowSG(storageGroupMNode);
+      mtree.init();
```

Comment:
Maybe we can put the `init()` inside the constructor of MTreeBelowSG ?

Reply:
Fixed. Now the init() is inside the constructor.